File: src/commands/shared/line-model.ts
const cLocaleDecoder = new TextDecoder('latin1');
const LINE_FEED = 10;

export interface ByteLine {
  bytes: Uint8Array;
  text: string;
}

export interface ByteLineModel {
  lines: ByteLine[];
  endsWithNewline: boolean;
  byteLength: number;
}

export function createByteLineModel(data: Uint8Array): ByteLineModel {
  const lines: ByteLine[] = [];
  let lineStart = 0;

  for (let index = 0; index < data.length; index += 1) {
    if (data[index] !== LINE_FEED) {
      continue;
    }

    const lineBytes = data.slice(lineStart, index);
    lines.push({
      bytes: lineBytes,
      text: decodeCLocaleText(lineBytes),
    });
    lineStart = index + 1;
  }

  const endsWithNewline = data.length > 0 && data[data.length - 1] === LINE_FEED;
  if (lineStart < data.length) {
    const lineBytes = data.slice(lineStart);
    lines.push({
      bytes: lineBytes,
      text: decodeCLocaleText(lineBytes),
    });
  }

  return {
    lines,
    endsWithNewline,
    byteLength: data.length,
  };
}

export function decodeCLocaleText(data: Uint8Array): string {
  return cLocaleDecoder.decode(data);
}

export function encodeCLocaleText(text: string): Uint8Array {
  const bytes = new Uint8Array(text.length);

  for (let index = 0; index < text.length; index += 1) {
    bytes[index] = text.charCodeAt(index) & 0xff;
  }

  return bytes;
}

File: python/src/onetool/commands/shared/line_model.py
from __future__ import annotations

from dataclasses import dataclass

LINE_FEED = 10


@dataclass(frozen=True, slots=True)
class ByteLine:
    bytes: bytes
    text: str


@dataclass(frozen=True, slots=True)
class ByteLineModel:
    lines: tuple[ByteLine, ...]
    ends_with_newline: bool
    byte_length: int


def create_byte_line_model(data: bytes) -> ByteLineModel:
    lines: list[ByteLine] = []
    line_start = 0

    for index, value in enumerate(data):
        if value != LINE_FEED:
            continue

        line_bytes = data[line_start:index]
        lines.append(ByteLine(bytes=line_bytes, text=decode_c_locale_text(line_bytes)))
        line_start = index + 1

    ends_with_newline = len(data) > 0 and data[-1] == LINE_FEED
    if line_start < len(data):
        line_bytes = data[line_start:]
        lines.append(ByteLine(bytes=line_bytes, text=decode_c_locale_text(line_bytes)))

    return ByteLineModel(
        lines=tuple(lines),
        ends_with_newline=ends_with_newline,
        byte_length=len(data),
    )


def decode_c_locale_text(data: bytes) -> str:
    return data.decode("latin1")


def encode_c_locale_text(text: str) -> bytes:
    return bytes(ord(char) & 0xFF for char in text)
